### PR TITLE
overlay: use private directory for composefs mounts

### PR DESCRIPTION
**Pull request by giuseppe**
similarly to what we are doing for the merged directory we need to use private directories for the composefs temporary mounts.  Otherwise when the id is in an additional store, we attempt to perform the mount without holding an exclusive lock there.

Could help with https://github.com/containers/storage/issues/2042.  Can't confirm as I wasn't able to find a reproducer yet.

@cgwalters PTAL, another fix for composefs

### CODE REVIEW

File: drivers/overlay/overlay.go [hunk 1]
```diff
@@ -1580,6 +1580,12 @@ func (d *Driver) get(id string, disableShifting bool, options graphdriver.MountO
 
 	composeFsLayers := []string{}
 	composeFsLayersDir := filepath.Join(dir, "composefs-layers")
+	if inAdditionalStore {
+		// Similarly to getMergedDir, we need to make sure we use a private directory for
+		// the composefs layers if in an additional store.  Otherwise, we could end up attempting
+		// a mount without having an exclusive lock on the additional store.
+		composeFsLayersDir = filepath.Join(d.runhome, id, "composefs-layers")
```
Comment (cgwalters) [inline drivers/overlay/overlay.go:1587]: There's a lot of hard to follow logic for me happening here. When I'm reviewing this PR I went to ask "ok what is `dir` here" - and it's the thing that gives us `inAdditionalStore`.

Then we have `mergedDir` which is computed using `inAdditionalStore` - and is seemingly written to solve *exactly* this issue, it says:

```
	// To prevent races with other processes mounting or unmounting the layer,
	// use a private directory under the main store rundir, not the "merged" 
```

So...would things be cleaner if we did something like:

```
var idDir string
if inAdditionalStore {
   idDir = path.Join(d.runhome, id)
} else {
   idDir = dir
}
```

There's probably a better name than `idDir`.

**Comment by rhatdan** [comment]
LGTM

**Review by cgwalters — COMMENTED** [review]
I can believe the patch fixes or helps with the reported bug, but I don't know this code well enough to just drop a lgtm myself yet, and I have some suggested cleanups.

File: drivers/overlay/overlay.go (hunk 1 above)
Comment (cgwalters) [reply drivers/overlay/overlay.go:1587]: Or maybe `getMergedDir()` changes to `getMountDir(..., "merged")` and `getMountDir(..., "composefs-layers")

Comment (giuseppe) [reply drivers/overlay/overlay.go]: thanks for the feedback.

Pushed a new version to refactor the `if inAdditionalStore { idDir = path.Join(d.runhome, id)} else {   idDir = dir }` in a separate function

**Comment by vrothberg** [comment]
@giuseppe lint is failing

**Comment by giuseppe** [comment]
rebased, it passes now

**Review by vrothberg — APPROVED** [review]
/lgtm
